library/core/src/tuple.rs

```diff
@@ -45,6 +46,27 @@ macro_rules! tuple_impls {
             {}
         }
 
+        maybe_tuple_doc! {
+            $($T)+ @
+            #[unstable(feature = "structural_match", issue = "31434")]
+            impl<$($T: ConstParamTy),+> ConstParamTy for ($($T,)+)
+            {}
+        }
+
+        maybe_tuple_doc! {
+            $($T)+ @
+            #[unstable(feature = "structural_match", issue = "31434")]
+            impl<$($T: StructuralPartialEq),+> StructuralPartialEq for ($($T,)+)
```

Comment:
This and `StructuralEq` do not need the `$T: Structural(Partial)Eq` bound since the trait is not intended to be recursive

Reply:
example, the following compiles even though `Foo` is not `StructuralEq` :
```rust
#![feature(structural_match)]

use core::marker::StructuralEq;

struct Foo;

impl Eq for Foo {}
impl PartialEq for Foo {
    fn eq(&self, other: &Foo) -> bool {
        true
    }
}

#[derive(Eq, PartialEq)]
struct Bar {
    a: Foo,
}

fn a<T: StructuralEq>() {}
fn b() {
    a::<Bar>();
}
```